sae_lens/toolkit/sae_patcher.py

```diff
@@ -0,0 +1,135 @@
+from typing import Literal, Protocol
+
+import torch
+from jaxtyping import Float
+from transformer_lens.hook_points import HookPoint
+
+from sae_lens import SparseAutoencoder
+
+
+class TransformerLensForwardHook(Protocol):
+    def __call__(self, orig: torch.Tensor, hook: HookPoint) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class TransformerLensBackwardHook(Protocol):
+    def __call__(self, orig: torch.Tensor, hook: HookPoint) -> tuple[torch.Tensor]:
+        raise NotImplementedError
+
+
+ForwardHookData = tuple[str, TransformerLensForwardHook]
+BackwardHookData = tuple[str, TransformerLensBackwardHook]
+
+NodeType = Literal["feature", "error", "all"]
+
+
+class SAEPatcher:
+    """Patches an SAE into the computational graph of a HookedTransformer
+
+    Usage:
+
+    sae_patcher = SAEPatcher(sae)
+    with model.hooks(
+        fwd_hooks=[sae_patcher.get_forward_hook()]
+        bwd_hooks=[sae_patcher.get_backward_hook()]
+    ):
+        ...
+    """
+
+    sae: SparseAutoencoder
+    sae_feature_acts: Float[torch.Tensor, "n_batch n_token d_sae"]
+    sae_errors: Float[torch.Tensor, "n_batch n_token d_model"]
+
+    def __init__(self, sae: SparseAutoencoder):
+        self.sae = sae
+        self.sae_feature_acts = torch.tensor([])
+        self.sae_errors = torch.tensor([])
+
+    def _forward_hook_fn(
+        self,
+        orig: Float[torch.Tensor, "n_batch n_token d_model"],
+        hook: HookPoint,
+    ) -> Float[torch.Tensor, "n_batch n_token d_model"]:
+        """Forward hook to patch the SAE into the computational graph
+
+        orig: Original activations
+        """
+
+        a_orig = orig
+        z_sae = self.sae.encode(a_orig)
+        # keep pyright happy
+        assert isinstance(z_sae, torch.Tensor)
+        a_sae = self.sae.decode(z_sae)
```

Comment:
I would worry about using a different forward pass calculation when patching, since it seems like it could result in subtle bugs if we change the implementation of `forward()` in the model. Can't we use the normal `.forward()` here, as it already returns the hidden activations of the SAE?